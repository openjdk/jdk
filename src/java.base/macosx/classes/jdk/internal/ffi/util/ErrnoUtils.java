/*
 * Copyright (c) 2024, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.ffi.util;

import jdk.internal.ffi.generated.errno.errno_h;

import java.io.IOException;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.nio.charset.StandardCharsets;

// Errno utility for macosx platform
public final class ErrnoUtils {
    private ErrnoUtils() {
    }

    private static final long ERRNO_STRING_HOLDER_ARRAY_SIZE = 256L;

    /**
     * Returns an IOException whose detail message combines the provided context with the
     * strerror_r description for the given errno.
     * On success, the message includes both the context and the errno string.
     * If strerror_r fails, the message only includes the context and errno value.
     * @param errno a non-negative POSIX errno value
     * @param context additional context to prefix the error message; must not be null
     * @return an IOException describing the errno
     */
    public static IOException ioExceptionFromErrno(int errno, String context) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment buf = arena.allocate(ERRNO_STRING_HOLDER_ARRAY_SIZE);
            if (errno_h.strerror_r(errno, buf, ERRNO_STRING_HOLDER_ARRAY_SIZE) == 0) {
                String errnoMsg = buf.getString(0, StandardCharsets.UTF_8);
                return new IOException(context + " " + errnoMsg);
            } else {
                // failed to convert errno to string - output errno value
                return new IOException(context + " Errno: " + errno);
            }
        }
    }
}
