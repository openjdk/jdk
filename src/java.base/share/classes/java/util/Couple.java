package java.util;

/**
 * The {@code Couple} class represents a mutable, ordered Couple of elements.
 * <p>
 * This class mimics the behavior of the C++ Standard Library's `std::Couple`, providing two values, referred to as
 * the "first" and the "second" elements, of possibly different types.
 * <p>
 * The {@code Couple} class provides a parameterized constructor to initialize both elements and allows access to them
 * through getter and setter methods.
 *
 * @param <K> the type of the first element
 * @param <V> the type of the second element
 *
 * @author Adarsh Sharma
 * @since 1.0
 */
public class Couple<K extends Comparable<? super K>, V extends Comparable<? super V>>
        implements Comparable<Couple<K, V>> {
    private K first;
    private V second;

    /**
     * Constructs a new Couple with the specified values.
     *
     * @param first the first element of the Couple
     * @param second the second element of the Couple
     */
    public Couple(K first, V second) {
        this.first = first;
        this.second = second;
    }

    /**
     * Returns the first element of the Couple.
     *
     * @return the first element of the Couple
     */
    public K getFirst() {
        return first;
    }

    /**
     * Sets the first element of the Couple.
     *
     * @param first the first element of the Couple
     */
    public void setFirst(K first) {
        this.first = first;
    }

    /**
     * Returns the second element of the Couple.
     *
     * @return the second element of the Couple
     */
    public V getSecond() {
        return second;
    }

    /**
     * Sets the second element of the Couple.
     *
     * @param second the second element of the Couple
     */
    public void setSecond(V second) {
        this.second = second;
    }

    /**
     * Compares this Couple with the specified Couple for order.
     * <p>
     * Comparison is based first on the {@code first} elements, and if they are equal,
     * then on the {@code second} elements. Both elements are compared using their natural ordering,
     * unless they are {@code null}, in which case {@code null} is considered less than non-null.
     * <p>
     * This method is {@code null}-safe and consistent with equals when both elements are non-null.
     *
     * @param other the Couple to be compared
     * @return a negative integer, zero, or a positive integer as this Couple
     *         is less than, equal to, or greater than the specified Couple
     * @throws NullPointerException if {@code other} is {@code null}
     */
    @Override
    public int compareTo(Couple<K, V> other) {
        if (other == null) {
            throw new NullPointerException("Compared Couple must not be null");
        }

        int cmp = compareNullable(this.first, other.first);
        if (cmp != 0) {
            return cmp;
        }

        return compareNullable(this.second, other.second);
    }

    /**
     * Helper method to compare two Comparable objects, handling nulls.
     * {@code null} is considered less than any non-null value.
     *
     * @param a first value
     * @param b second value
     * @return comparison result
     */
    @SuppressWarnings("unchecked")
    private <T extends Comparable<? super T>> int compareNullable(T a, T b) {
        if (a == b) return 0;
        if (a == null) return -1;
        if (b == null) return 1;
        return a.compareTo(b);
    }

    /**
     * Returns a string representation of this Couple in the format {@code (first, second)}.
     *
     * @return a string representation of this Couple
     */
    @Override
    public String toString() {
        return "(" + first + ", " + second + ")";
    }

    /**
     * Checks if this Couple is equal to another object. Two Couples are considered equal if both the first and second
     * elements are equal.
     *
     * @param o the object to compare this Couple with
     * @return {@code true} if this Couple is equal to the specified object
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Couple<?, ?> Couple = (Couple<?, ?>) o;
        return Objects.equals(first, Couple.first) && Objects.equals(second, Couple.second);
    }

    /**
     * Returns the hash code for this Couple, computed based on the hash codes of the first and second elements.
     *
     * @return the hash code for this Couple
     */
    @Override
    public int hashCode() {
        return Objects.hash(first, second);
    }

    /**
     * Creates a new {@code Couple} with the given first and second elements.
     *
     * @param first the first element of the Couple
     * @param second the second element of the Couple
     * @param <K> the type of the first element, must be Comparable
     * @param <V> the type of the second element, must be Comparable
     * @return a new {@code Couple} containing the given elements
     */
    public static <K extends Comparable<? super K>, V extends Comparable<? super V>>
    Couple<K, V> makeCouple(K first, V second) {
        return new Couple<>(first, second);
    }
}
