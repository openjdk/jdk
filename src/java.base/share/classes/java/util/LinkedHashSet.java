/*
 * Copyright (c) 2000, 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.util;

/**
 * <p>Hash table and linked list implementation of the {@code Set} interface,
 * with well-defined encounter order.  This implementation differs from
 * {@code HashSet} in that it maintains a doubly-linked list running through
 * all of its entries.  This linked list defines the encounter order (iteration
 * order), which is the order in which elements were inserted into the set
 * (<i>insertion-order</i>). The least recently inserted element (the eldest) is
 * first, and the youngest element is last. Note that encounter order is <i>not</i> affected
 * if an element is <i>re-inserted</i> into the set with the {@code add} method.
 * (An element {@code e} is reinserted into a set {@code s} if {@code s.add(e)} is
 * invoked when {@code s.contains(e)} would return {@code true} immediately prior to
 * the invocation.) The reverse-ordered view of this set is in the opposite order, with
 * the youngest element appearing first and the eldest element appearing last. The encounter
 * order of elements already in the set can be changed by using the
 * {@link #addFirst addFirst} and {@link #addLast addLast} methods.
 *
 * <p>This implementation spares its clients from the unspecified, generally
 * chaotic ordering provided by {@link HashSet}, without incurring the
 * increased cost associated with {@link TreeSet}.  It can be used to
 * produce a copy of a set that has the same order as the original, regardless
 * of the original set's implementation:
 * <pre>{@code
 *     void foo(Set<String> s) {
 *         Set<String> copy = new LinkedHashSet<>(s);
 *         ...
 *     }
 * }</pre>
 * This technique is particularly useful if a module takes a set on input,
 * copies it, and later returns results whose order is determined by that of
 * the copy.  (Clients generally appreciate having things returned in the same
 * order they were presented.)
 *
 * <p>This class provides all of the optional {@link Set} and {@link SequencedSet}
 * operations, and it permits null elements. Like {@code HashSet}, it provides constant-time
 * performance for the basic operations ({@code add}, {@code contains} and
 * {@code remove}), assuming the hash function disperses elements
 * properly among the buckets.  Performance is likely to be just slightly
 * below that of {@code HashSet}, due to the added expense of maintaining the
 * linked list, with one exception: Iteration over a {@code LinkedHashSet}
 * requires time proportional to the <i>size</i> of the set, regardless of
 * its capacity.  Iteration over a {@code HashSet} is likely to be more
 * expensive, requiring time proportional to its <i>capacity</i>.
 *
 * <p>A linked hash set has two parameters that affect its performance:
 * <i>initial capacity</i> and <i>load factor</i>.  They are defined precisely
 * as for {@code HashSet}.  Note, however, that the penalty for choosing an
 * excessively high value for initial capacity is less severe for this class
 * than for {@code HashSet}, as iteration times for this class are unaffected
 * by capacity.
 *
 * <p><strong>Note that this implementation is not synchronized.</strong>
 * If multiple threads access a linked hash set concurrently, and at least
 * one of the threads modifies the set, it <em>must</em> be synchronized
 * externally.  This is typically accomplished by synchronizing on some
 * object that naturally encapsulates the set.
 *
 * If no such object exists, the set should be "wrapped" using the
 * {@link Collections#synchronizedSet Collections.synchronizedSet}
 * method.  This is best done at creation time, to prevent accidental
 * unsynchronized access to the set: <pre>
 *   Set s = Collections.synchronizedSet(new LinkedHashSet(...));</pre>
 *
 * <p>The iterators returned by this class's {@code iterator} method are
 * <em>fail-fast</em>: if the set is modified at any time after the iterator
 * is created, in any way except through the iterator's own {@code remove}
 * method, the iterator will throw a {@link ConcurrentModificationException}.
 * Thus, in the face of concurrent modification, the iterator fails quickly
 * and cleanly, rather than risking arbitrary, non-deterministic behavior at
 * an undetermined time in the future.
 *
 * <p>Note that the fail-fast behavior of an iterator cannot be guaranteed
 * as it is, generally speaking, impossible to make any hard guarantees in the
 * presence of unsynchronized concurrent modification.  Fail-fast iterators
 * throw {@code ConcurrentModificationException} on a best-effort basis.
 * Therefore, it would be wrong to write a program that depended on this
 * exception for its correctness:   <i>the fail-fast behavior of iterators
 * should be used only to detect bugs.</i>
 *
 * <p>This class is a member of the
 * <a href="{@docRoot}/java.base/java/util/package-summary.html#CollectionsFramework">
 * Java Collections Framework</a>.
 *
 * @param <E> the type of elements maintained by this set
 *
 * @author  Josh Bloch
 * @see     Object#hashCode()
 * @see     Collection
 * @see     Set
 * @see     HashSet
 * @see     TreeSet
 * @see     Hashtable
 * @since   1.4
 */

public class LinkedHashSet<E>
    extends HashSet<E>
    implements SequencedSet<E>, Cloneable, java.io.Serializable {

    @java.io.Serial
    private static final long serialVersionUID = -2851667679971038690L;

    /**
     * Constructs a new, empty linked hash set with the specified initial
     * capacity and load factor.
     *
     * @apiNote
     * To create a {@code LinkedHashSet} with an initial capacity that accommodates
     * an expected number of elements, use {@link #newLinkedHashSet(int) newLinkedHashSet}.
     *
     * @param      initialCapacity the initial capacity of the linked hash set
     * @param      loadFactor      the load factor of the linked hash set
     * @throws     IllegalArgumentException  if the initial capacity is less
     *               than zero, or if the load factor is nonpositive
     */
    public LinkedHashSet(int initialCapacity, float loadFactor) {
        super(initialCapacity, loadFactor, true);
    }

    /**
     * Constructs a new, empty linked hash set with the specified initial
     * capacity and the default load factor (0.75).
     *
     * @apiNote
     * To create a {@code LinkedHashSet} with an initial capacity that accommodates
     * an expected number of elements, use {@link #newLinkedHashSet(int) newLinkedHashSet}.
     *
     * @param   initialCapacity   the initial capacity of the LinkedHashSet
     * @throws  IllegalArgumentException if the initial capacity is less
     *              than zero
     */
    public LinkedHashSet(int initialCapacity) {
        super(initialCapacity, .75f, true);
    }

    /**
     * Constructs a new, empty linked hash set with the default initial
     * capacity (16) and load factor (0.75).
     */
    public LinkedHashSet() {
        super(16, .75f, true);
    }

    /**
     * Constructs a new linked hash set with the same elements as the
     * specified collection.  The linked hash set is created with an initial
     * capacity sufficient to hold the elements in the specified collection
     * and the default load factor (0.75).
     *
     * @param c  the collection whose elements are to be placed into
     *           this set
     * @throws NullPointerException if the specified collection is null
     */
    @SuppressWarnings("this-escape")
    public LinkedHashSet(Collection<? extends E> c) {
        super(HashMap.calculateHashMapCapacity(Math.max(c.size(), 12)), .75f, true);
        addAll(c);
    }

    /**
     * Creates a <em><a href="Spliterator.html#binding">late-binding</a></em>
     * and <em>fail-fast</em> {@code Spliterator} over the elements in this set.
     *
     * <p>The {@code Spliterator} reports {@link Spliterator#SIZED},
     * {@link Spliterator#DISTINCT}, and {@code ORDERED}.  Implementations
     * should document the reporting of additional characteristic values.
     *
     * @implNote
     * The implementation creates a
     * <em><a href="Spliterator.html#binding">late-binding</a></em> spliterator
     * from the set's {@code Iterator}.  The spliterator inherits the
     * <em>fail-fast</em> properties of the set's iterator.
     * The created {@code Spliterator} additionally reports
     * {@link Spliterator#SUBSIZED}.
     *
     * @return a {@code Spliterator} over the elements in this set
     * @since 1.8
     */
    @Override
    public Spliterator<E> spliterator() {
        return Spliterators.spliterator(this, Spliterator.DISTINCT | Spliterator.ORDERED);
    }

    /**
     * Creates a new, empty LinkedHashSet suitable for the expected number of elements.
     * The returned set uses the default load factor of 0.75, and its initial capacity is
     * generally large enough so that the expected number of elements can be added
     * without resizing the set.
     *
     * @param numElements    the expected number of elements
     * @param <T>         the type of elements maintained by the new set
     * @return the newly created set
     * @throws IllegalArgumentException if numElements is negative
     * @since 19
     */
    public static <T> LinkedHashSet<T> newLinkedHashSet(int numElements) {
        if (numElements < 0) {
            throw new IllegalArgumentException("Negative number of elements: " + numElements);
        }
        return new LinkedHashSet<>(HashMap.calculateHashMapCapacity(numElements));
    }

    @SuppressWarnings("unchecked")
    LinkedHashMap<E, Object> map() {
        return (LinkedHashMap<E, Object>) map;
    }

    /**
     * {@inheritDoc}
     * <p>
     * If this set already contains the element, it is relocated if necessary so that it is
     * first in encounter order.
     *
     * @since 21
     */
    public void addFirst(E e) {
        map().putFirst(e, PRESENT);
    }

    /**
     * {@inheritDoc}
     * <p>
     * If this set already contains the element, it is relocated if necessary so that it is
     * last in encounter order.
     *
     * @since 21
     */
    public void addLast(E e) {
        map().putLast(e, PRESENT);
    }

    /**
     * {@inheritDoc}
     *
     * @throws NoSuchElementException {@inheritDoc}
     * @since 21
     */
    public E getFirst() {
        return map().sequencedKeySet().getFirst();
    }

    /**
     * {@inheritDoc}
     *
     * @throws NoSuchElementException {@inheritDoc}
     * @since 21
     */
    public E getLast() {
        return map().sequencedKeySet().getLast();
    }

    /**
     * {@inheritDoc}
     *
     * @throws NoSuchElementException {@inheritDoc}
     * @since 21
     */
    public E removeFirst() {
        return map().sequencedKeySet().removeFirst();
    }

    /**
     * {@inheritDoc}
     *
     * @throws NoSuchElementException {@inheritDoc}
     * @since 21
     */
    public E removeLast() {
        return map().sequencedKeySet().removeLast();
    }

    /**
     * {@inheritDoc}
     * <p>
     * Modifications to the reversed view are permitted and will be propagated to this set.
     * In addition, modifications to this set will be visible in the reversed view.
     *
     * @return {@inheritDoc}
     * @since 21
     */
    public SequencedSet<E> reversed() {
        class ReverseLinkedHashSetView extends AbstractSet<E> implements SequencedSet<E> {
            public int size()                  { return LinkedHashSet.this.size(); }
            public Iterator<E> iterator()      { return map().sequencedKeySet().reversed().iterator(); }
            public boolean add(E e)            { return LinkedHashSet.this.add(e); }
            public void addFirst(E e)          { LinkedHashSet.this.addLast(e); }
            public void addLast(E e)           { LinkedHashSet.this.addFirst(e); }
            public E getFirst()                { return LinkedHashSet.this.getLast(); }
            public E getLast()                 { return LinkedHashSet.this.getFirst(); }
            public E removeFirst()             { return LinkedHashSet.this.removeLast(); }
            public E removeLast()              { return LinkedHashSet.this.removeFirst(); }
            public SequencedSet<E> reversed()  { return LinkedHashSet.this; }
            public Object[] toArray() { return map().keysToArray(new Object[map.size()], true); }
            public <T> T[] toArray(T[] a) { return map().keysToArray(map.prepareArray(a), true); }
        }

        return new ReverseLinkedHashSetView();
    }
}
