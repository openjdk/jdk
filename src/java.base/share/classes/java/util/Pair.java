package java.util;

/**
 * The {@code Pair} class represents a mutable, ordered pair of elements.
 * <p>
 * This class mimics the behavior of the C++ Standard Library's `std::pair`, providing two values, referred to as
 * the "first" and the "second" elements, of possibly different types.
 * <p>
 * The {@code Pair} class provides a parameterized constructor to initialize both elements and allows access to them
 * through getter and setter methods.
 *
 * @param <K> the type of the first element
 * @param <V> the type of the second element
 *
 * @author Adarsh Sharma
 * @since 1.0
 */
public class Pair<K, V> {
    private K first;
    private V second;

    /**
     * Constructs a new pair with the specified values.
     *
     * @param first the first element of the pair
     * @param second the second element of the pair
     */
    public Pair(K first, V second) {
        this.first = first;
        this.second = second;
    }

    /**
     * Returns the first element of the pair.
     *
     * @return the first element of the pair
     */
    public K getFirst() {
        return first;
    }

    /**
     * Sets the first element of the pair.
     *
     * @param first the first element of the pair
     */
    public void setFirst(K first) {
        this.first = first;
    }

    /**
     * Returns the second element of the pair.
     *
     * @return the second element of the pair
     */
    public V getSecond() {
        return second;
    }

    /**
     * Sets the second element of the pair.
     *
     * @param second the second element of the pair
     */
    public void setSecond(V second) {
        this.second = second;
    }

    /**
     * Swaps the first and second elements of the pair.
     */
    public void swap() {
        K temp = first;
        first = second;
        second = temp;
    }

    /**
     * Compares this pair with the specified pair for order.
     * <p>
     * Comparison is based first on the {@code first} elements, and if they are equal,
     * then on the {@code second} elements. Both elements are compared using their natural ordering,
     * unless they are {@code null}, in which case {@code null} is considered less than non-null.
     * <p>
     * This method is {@code null}-safe and consistent with equals when both elements are non-null.
     *
     * @param other the pair to be compared
     * @return a negative integer, zero, or a positive integer as this pair
     *         is less than, equal to, or greater than the specified pair
     * @throws NullPointerException if {@code other} is {@code null}
     */
    @Override
    public int compareTo(Pair<K, V> other) {
        if (other == null) {
            throw new NullPointerException("Compared pair must not be null");
        }

        int cmp = compareNullable(this.first, other.first);
        if (cmp != 0) {
            return cmp;
        }

        return compareNullable(this.second, other.second);
    }

    /**
     * Helper method to compare two Comparable objects, handling nulls.
     * {@code null} is considered less than any non-null value.
     *
     * @param a first value
     * @param b second value
     * @return comparison result
     */
    @SuppressWarnings("unchecked")
    private <T extends Comparable<? super T>> int compareNullable(T a, T b) {
        if (a == b) return 0;
        if (a == null) return -1;
        if (b == null) return 1;
        return a.compareTo(b);
    }

    /**
     * Returns a string representation of this pair in the format {@code (first, second)}.
     *
     * @return a string representation of this pair
     */
    @Override
    public String toString() {
        return "(" + String.valueOf(first) + ", " + String.valueOf(second) + ")";
    }

    /**
     * Checks if this pair is equal to another object. Two pairs are considered equal if both the first and second
     * elements are equal.
     *
     * @param o the object to compare this pair with
     * @return {@code true} if this pair is equal to the specified object
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pair<?, ?> pair = (Pair<?, ?>) o;
        return first.equals(pair.first) && second.equals(pair.second);
    }

    /**
     * Returns the hash code for this pair, computed based on the hash codes of the first and second elements.
     *
     * @return the hash code for this pair
     */
    @Override
    public int hashCode() {
        return Objects.hash(first, second);
    }

    /**
     * Creates a new {@code Pair} with the given first and second elements.
     *
     * @param first the first element of the pair
     * @param second the second element of the pair
     * @param <K> the type of the first element
     * @param <V> the type of the second element
     * @return a new {@code Pair} containing the given elements
     */
    public static <K, V> Pair<K, V> makePair(K first, V second) {
        return new Pair<>(first, second);
    }
}
