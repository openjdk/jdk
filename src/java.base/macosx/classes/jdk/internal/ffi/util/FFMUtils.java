/*
 * Copyright (c) 2024, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.ffi.util;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.util.Arrays;
import java.util.stream.Collectors;

import jdk.internal.misc.Unsafe;

@SuppressWarnings("restricted")
public final class FFMUtils {

    public static final ValueLayout.OfBoolean C_BOOL =
            (ValueLayout.OfBoolean) Linker.nativeLinker().canonicalLayouts().get("bool");

    public static final ValueLayout.OfByte C_CHAR =
            (ValueLayout.OfByte)Linker.nativeLinker().canonicalLayouts().get("char");

    public static final ValueLayout.OfShort C_SHORT =
            (ValueLayout.OfShort) Linker.nativeLinker().canonicalLayouts().get("short");

    public static final ValueLayout.OfInt C_INT =
            (ValueLayout.OfInt) Linker.nativeLinker().canonicalLayouts().get("int");

    public static final ValueLayout.OfLong C_LONG_LONG =
            (ValueLayout.OfLong) Linker.nativeLinker().canonicalLayouts().get("long long");

    public static final ValueLayout.OfFloat C_FLOAT =
            (ValueLayout.OfFloat) Linker.nativeLinker().canonicalLayouts().get("float");

    public static final ValueLayout.OfDouble C_DOUBLE =
            (ValueLayout.OfDouble) Linker.nativeLinker().canonicalLayouts().get("double");

    public static final AddressLayout C_POINTER =
            ((AddressLayout) Linker.nativeLinker().canonicalLayouts().get("void*"))
                    .withTargetLayout(MemoryLayout.sequenceLayout(Long.MAX_VALUE, C_CHAR));

    public static final ValueLayout.OfLong C_LONG =
            (ValueLayout.OfLong) Linker.nativeLinker().canonicalLayouts().get("long");

    private FFMUtils() {
    }

    /**
     * Returns a {@code MemorySegment} set to the size of byteSize
     *
     * @param byteSize the size in bytes to be allocated
     * @param byteAlignment the size in bytes for the memory alignment
     *
     * @throws IllegalArgumentException if the maxByteAlignment of the created
     * MemorySegment is less than the provided byteAlignment
     *
     * @return the newly created {@code MemorySegment}
     */
    public static MemorySegment malloc(long byteSize, long byteAlignment) {
        long allocatedMemory = UNSAFE.allocateMemory(byteSize);
        MemorySegment result = MemorySegment.ofAddress(allocatedMemory).reinterpret(byteSize);
        if (result.maxByteAlignment() < byteAlignment) {
            throw new IllegalArgumentException();
        }
        return result;
    }

    /**
     * Takes a {@code MemorySegment} and deallocates the memory at that address
     * @param memorySegment the {@code MemorySegment} that will be deallocated
     */
    public static void free(MemorySegment memorySegment) {
        UNSAFE.freeMemory(memorySegment.address());
    }

    // SegmentAllocator that delegates to Unsafe for memory allocation
    public static final SegmentAllocator SEGMENT_ALLOCATOR = new SegmentAllocator() {
        @Override
        public MemorySegment allocate(long byteSize, long byteAlignment) {
            return malloc(byteSize, byteAlignment);
        }
    };

    private static final Unsafe UNSAFE = Unsafe.getUnsafe();

    // Variables and methods below are extracted from jextract generated
    // code and used by native bindings on all platforms
    public static final boolean TRACE_DOWNCALLS = false;
    static final SymbolLookup SYMBOL_LOOKUP = SymbolLookup.loaderLookup()
            .or(Linker.nativeLinker().defaultLookup());

    public static void traceDowncall(String name, Object... args) {
        String traceArgs = Arrays.stream(args)
                .map(Object::toString)
                .collect(Collectors.joining(", "));
        System.out.printf("%s(%s)\n", name, traceArgs);
    }

    public static MemorySegment findOrThrow(String symbol) {
        return SYMBOL_LOOKUP.findOrThrow(symbol);
    }

    public static MethodHandle upcallHandle(Class<?> fi, String name, FunctionDescriptor fdesc) {
        try {
            return MethodHandles.lookup().findVirtual(fi, name, fdesc.toMethodType());
        } catch (ReflectiveOperationException ex) {
            throw new AssertionError(ex);
        }
    }

    public static MemoryLayout align(MemoryLayout layout, long align) {
        return switch (layout) {
            case PaddingLayout p -> p;
            case ValueLayout v -> v.withByteAlignment(align);
            case GroupLayout g -> {
                MemoryLayout[] alignedMembers = g.memberLayouts().stream()
                        .map(m -> align(m, align)).toArray(MemoryLayout[]::new);
                yield g instanceof StructLayout ?
                        MemoryLayout.structLayout(alignedMembers) : MemoryLayout.unionLayout(alignedMembers);
            }
            case SequenceLayout s -> MemoryLayout.sequenceLayout(s.elementCount(), align(s.elementLayout(), align));
        };
    }
}
