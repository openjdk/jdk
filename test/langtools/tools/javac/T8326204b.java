import java.util.Map;

/*
 * @test /nodynamiccopyright/
 * @bug 8326204
 * @summary yield statements doesn't allow cast expressions with more than 1 type arguments
 * @compile/fail/ref=T8326204b.out -XDrawDiagnostics --should-stop=at=FLOW -XDdev T8326204b.java
 */
public class T8326204b {
    private static void t(int i) { yield((Map<String, String>) null, 2); }

    private static void yield(Map<String, String> m, int j) { }
}
