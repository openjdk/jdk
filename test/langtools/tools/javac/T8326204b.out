T8326204b.java:10:36: compiler.err.invalid.yield
1 error
